packages/@lwc/template-compiler/src/parser/index.ts

```diff
@@ -976,24 +974,26 @@ function validateTemplate(
     parse5Elm: parse5.Element,
     parse5ElmLocation: parse5.ElementLocation
 ): void {
-    if (parse5Elm.tagName === 'template') {
-        const location = ast.sourceLocation(parse5ElmLocation);
+    const location = ast.sourceLocation(parse5ElmLocation);
```

Comment:
I wouldn't mind seeing a guard at the top of this function. It doesn't fix an existing problem, but does solve a future problem: where one could call `validateTemplate` with a value for `parse5Elm` that is not a `template` tag.

### The TypeScript Way:

Introduce a new type:
```typescript
type TemplateElement = parse5.Element & { tagName: 'template' }
```
and change the type of `parse5Elm` above to `TemplateElement`. Elsewhere, when passing in a value for `parse5Elm` (like in `parseElement`), you'll have to explicitly write `as TemplateElement`, which will make it harder to do the wrong thing.

### The JavaScript way

Guard the function at runtime, with something like:

```typescript
if (parse5Elm.tagName === 'template') {
    threw new Error('validateTemplate was called with an element that is not a template.');
}
```

This should never throw. But if we introduce a bug sometime in the future, it'll make things very obvious what we did wrong. And, since the `parse5Elm.tagName` will always be `template`, the `if` block would be optimized away by the JIT at runtime resulting in zero performance impact.

I think the TypeScript way is probably more consistent with the rest of the LWC codebase (and is therefore what I'd recommend), although the JavaScript way provides a stronger guarantee.

Reply:
Great suggestion!